torch/quantization/fx/_equalize.py

```diff
@@ -220,3 +222,245 @@ def node_supports_equalization(node: Node, modules) -> bool:
+
+    After:
+                                              scaled weight values
+                                                      |
+                                                 WeightEqObs
+                                                      |
+       equalization scale                       WeightQuantObs
+              |                                       |
+        x -> mul -> InpQuantObs (scaled min/max) -> linear -> OutQuantObs
+    """
+    for node in model.graph.nodes:
+        if node.op == 'call_module' and isinstance(modules[node.target], _InputEqualizationObserver):
+            inp_quant_obs_node = node.args[0]
+            prev_node = inp_quant_obs_node.args[0]
+
+            # TODO: Possible special handling for connected linear layers
+            # Update the following input quantization observer's min/max values
+            scale_input_node(node, modules)
+
+            # Remove the InputEqualization node and add a mul operator before
+            # the quantization observer node that appears before the equalization node
+            # Before: x -> input_quant_obs -> input_eq_obs -> linear
+            # After: x -> mul -> input_quant_obs -> linear
+
+            # Create a node containing the equalization scale
+            with model.graph.inserting_before(inp_quant_obs_node):
+                name = node.name + '_scale'
+                setattr(model, name, modules[node.target].equalization_scale)
+                eq_scale_node = model.graph.create_node('get_attr', name)
+
+            # Create a node multiplying the input with the equalization scale
+            with model.graph.inserting_after(eq_scale_node):
+                inputs = (prev_node, eq_scale_node)
+                mul_node = model.graph.create_node("call_function", torch.mul, inputs)
+
+            # Set the mul nod to be the input_quant_obs_node's input instead of
+            # the previous node
+            inp_quant_obs_node.replace_input_with(prev_node, mul_node)
+
+            # For all of the current node's users, replace the current node with
+            # the input quantization observer node
+            orig_users = list(node.users.keys())
+            for user_node in orig_users:
+                user_node.replace_input_with(node, inp_quant_obs_node)
+
+            # Erase the InputEqualizationObserver node
+            model.graph.erase_node(node)
+
+        elif weight_eq_obs_dict.get(node.name, None) is not None:
+            weight_eq_obs = weight_eq_obs_dict.get(node.name)
+            assert(isinstance(weight_eq_obs, _WeightEqualizationObserver))
+            assert(isinstance(node.target, str))
+
+            equalization_scale = weight_eq_obs.equalization_scale
+
+            # Scales the weights and runs the weight quantization observers
+            next_equalization_scale = get_next_equalization_scale(node, modules)
+            scale_weight_node(node, modules, equalization_scale, next_equalization_scale)
```

Comment:
Where do we run the weight quant observer after scaling the weight node?

Reply:
I noticed that instead of running the weight quant observer we can just edit the weight node itself. My understanding is that at this point, the weight quantization observer has not been created yet, so later in `convert` when `from_float` gets called, then the weight quantization observer will be created and run with the scaled weights. Let me know if this is not the case.

Reply:
For `F.linear` layers, I noticed that the weight quantization observer gets rerun when [fold_weight(https://github.com/pytorch/pytorch/blob/master/torch/quantization/fx/convert.py#L74) gets called at the very end so if we update the weight value, then this rerun will also update the qparams in the weight quantization observer.

Reply:
Do you mean https://github.com/pytorch/pytorch/blob/1d5a577f04fbd67824e7b74aaa1ce3ec72f5557d/torch/quantization/fx/convert.py#L159? I think we need to confirm that the plan is to run the weight observer for all cases static/dynamic, nn.Linear/F.Linear. @jerryzh168 can you confirm?

Reply:
yeah I think so

Reply:
Is the plan in the future to run all weight observers (for both `nn.Linear` and `F.linear`) at the beginning of `convert`? Because it seems like right now that `run_weight_observers` only runs the observers for `call_function` operation nodes. 

For `nn.Linear` layers, it seems like they are currently being run in [from_float](https://github.com/pytorch/pytorch/blob/master/torch/nn/quantized/modules/linear.py#L269). For `F.linear` layers, it seems like they are currently being re-run in [fold_weight](https://github.com/pytorch/pytorch/blob/master/torch/quantization/fx/convert.py#L95) being called at the very end [here](https://github.com/pytorch/pytorch/blob/master/torch/quantization/fx/convert.py#L480)

Reply:
yeah for nn.Linear we are running weight_observer when we instantiate that in from_float so there is no way to run in the beginning currently.

fold_weight is not running observers I thnk, it's running the quantization ops for weight, e.g. quantize_per_tensor/quantize_per_channel0

Reply:
Oh, ok, then I will probably need to edit the code for functional layers. Would it be possible move the call to `run_weight_observers` to after the call to equalization functions? Currently, `run_weight_observers` is called before the call to `find_matches`. But I think I need to call the equalization functions after `find_matches` since the nodes inserted in the equalization functions will not have a `qconfig_map` entry and will cause an error in `find_matches`.

Reply:
sounds good, also it would be good to have a higher level descriptions of things we are doing in the PR as well so that it's easier to understand how it will interact with the other parts. The detailed explanation of each step looks good.